arborist/server.go

```diff
@@ -1304,13 +1304,43 @@ func (server *Server) handleUserRevokeAll(w http.ResponseWriter, r *http.Request
 func (server *Server) handleUserRevokePolicy(w http.ResponseWriter, r *http.Request) {
 	username := mux.Vars(r)["username"]
 	policyName := mux.Vars(r)["policyName"]
-	errResponse := revokeUserPolicy(server.db, username, policyName, getAuthZProvider(r))
-	if errResponse != nil {
-		errResponse.log.write(server.logger)
-		_ = errResponse.write(w, r)
-		return
+	authzProvider := getAuthZProvider(r)
+	policyInfo, _ := fetchPolicyInfo(server.db, username, policyName)
+
+	if policyInfo != nil {
+		server.logger.Info("Policy - {name: %s, authz_provider: %s, type: %s}",
+			policyInfo.Name, policyInfo.Authz_provider, policyInfo.Type)
+
+		if policyInfo.Type == "group" {
+			msg := fmt.Sprintf(
+				"Policy `%s` is assigned to the user through a group. Can not be revoked directly",
+				policyName)
+			errResponse := newErrorResponse(msg, http.StatusBadRequest, nil)
+			errResponse.log.write(server.logger)
+			_ = errResponse.write(w, r)
+			return
+		}
+
+		if policyInfo.Authz_provider == authzProvider.String {
+			errResponse := revokeUserPolicy(
+				server.db, username, policyName, authzProvider)
+			if errResponse != nil {
+				errResponse.log.write(server.logger)
+				_ = errResponse.write(w, r)
+				return
+			}
+			server.logger.Info("revoked policy %s for user %s", policyName, username)
+		} else {
+			msg := fmt.Sprintf("Cannot revoke policy `%s`.Authz_provider - `%s` and `%s` Mismatch",
```

Comment:
i think you should definitely log this, but not return it to the user - telling the user "authz_provider mismatch" without the provider names is enough information imo